sympy/matrices/matrices.py

```diff
@@ -3825,8 +3825,11 @@ def pinv(self):
                 return AH * (A * AH).inv()
         except ValueError:
             # Matrix is not full rank, so A*AH cannot be inverted.
-            raise NotImplementedError('Rank-deficient matrices are not yet '
-                                      'supported.')
+            P, D = (A.H * A).diagonalize()
```

Comment:
`A.H` has already been computed as `AH`. Let's reuse `AH`.

The `.diagonalize()` function might throw an exception if it cannot find the eigenvalues. Let's put this in a try except block, catching the correct exception.

Reply:
Just checked diagonalize() and it appears it can only throw a MatrixError if is_diagonalizable() returns false. Since (AH * A) is Hermitian that cannot happen: (AH * A) is guaranteed to be diagonalizable with real values on its diagonal. Or rather, if it throws, that just means there is a bug in is_diagonalizable(). We shouldn't compensate for that should we?

I'll reuse AH.

Reply:
Yes, it's mathematically diagonalizable, but that doesn't mean that sympy will succeed in factoring the characteristic polynomial. Try diagonalizing a random 5x5 with sympy.

Reply:
Fair enough. Still, it means that pinv() fails even though it should succeed. In turn that means that we should still throw an exception - we can't return a result after all. We can just let the existing exception propagate, can't we?

Reply:
I think we should raise a `ValueError` like before that says something like `pinv for matrices where the characteristic polynomial of A*A.H is not factorable are not yet supported.` Do you know a different algorithm that could be used in that case?

Reply:
The alternative I know is SVD, but I think that will only be worse, even if it was available in sympy.
Ok, I'll catch MatrixError and raise a ValueError as suggested.

Reply:
That should be a NotImplementedError instead of a ValueError, like it was before, shouldn't it?